### PR TITLE
Stabilize `transmute` in constants and statics but not const fn

### PR DESCRIPTION
**Pull request by oli-obk**
cc #53605 (leaving issue open so we can add `transmute` to `const fn` later)

Previous attempt: #64011 

r? @RalfJung 

cc @rust-lang/wg-const-eval

### CODE REVIEW

File: src/librustc_mir/transform/qualify_min_const_fn.rs [hunk 1]
```diff
@@ -382,6 +383,15 @@ fn check_terminator(
                         .into(),
                     ));
                 }
+                if tcx.fn_sig(def_id).abi() == RustIntrinsic
+                    && tcx.item_name(def_id) == sym::transmute
+                    && !feature_allowed(tcx, def_id, sym::transmute_in_const_fn)
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_min_const_fn.rs:388]: This is the fun part. It works, but it's absolutely not scalable to more functions or intrinsics.

**Comment by oli-obk** [comment]
Idk who needs to sign off on this stabilization. Technically T-libs? Hypothetically just wg-const-eval? I mean this is just removing boilerplate for users for a feature already existing via `union`s.

File: src/librustc_feature/active.rs [hunk 2]
```diff
@@ -574,6 +574,9 @@ declare_features! (
     /// No longer treat an unsafe function as an unsafe block.
     (active, unsafe_block_in_unsafe_fn, "1.45.0", Some(71668), None),
 
+    /// Alloc calling `transmute` in const fn
+    (active, transmute_in_const_fn, "1.45.0", Some(53605), None),
```
Comment (ecstatic-morse) [inline src/librustc_feature/active.rs:578]: I've been trying to prefix all lang and lib features pertaining to const-eval with "const". Could we continue that by calling this `const_fn_transmute` or so?

File: src/librustc_mir/transform/qualify_min_const_fn.rs [hunk 3]
```diff
@@ -382,6 +383,15 @@ fn check_terminator(
...
+                {
+                    return Err((
+                        span,
+                        "can only call `transmute` from const items, not `const fn`".into(),
+                    ));
+                }
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs:394]: I noticed that, but I don't know what our `const_fn` feature gate story is anymore. iirc it's "the `const_fn` feature will go away when we merge `min_const_fn` into the regular const checks"?

[the remark this answers is not in the document]
Comment (ecstatic-morse) [reply src/librustc_mir/transform/qualify_min_const_fn.rs:394]: I think so. I believe it's only blocked on me finding the willpower to update several hundred `//~ ERROR` annotations in the `min_const_fn` suite. Centril had some objections IIRC, but I don't think they were substantive.

In the meantime, we should check this in `check_consts/validation.rs` as well. Adding a structured error type for `transmute` and checking for it where we handle `TerminatorKind::Call` should be enough.

File: src/libcore/intrinsics.rs [hunk 4]
```diff
@@ -1284,7 +1284,7 @@ extern "rust-intrinsic" {
     /// }
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
-    #[rustc_const_unstable(feature = "const_transmute", issue = "53605")]
+    #[rustc_const_stable(feature = "const_transmute", since = "1.45.0")]
```
Comment (RalfJung) [inline src/libcore/intrinsics.rs:1287]: Maybe add a comment that this is actually only stable in some contexts?

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/transform/qualify_min_const_fn.rs:388]: Maybe add a comment `// HACK` or so, just to document that point.

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/transform/qualify_min_const_fn.rs:394]: Also we should have good test coverage for this one to ensure we don't accidentally stabilize `transmute` in `const fn` under strange circumstances...

File: src/test/ui/feature-gates/feature-gate-const_transmute.rs [hunk 5]
```diff
@@ -4,6 +4,8 @@ use std::mem;
 struct Foo(u32);
 
 const TRANSMUTED_U32: u32 = unsafe { mem::transmute(Foo(3)) };
-//~^ ERROR `std::intrinsics::transmute` is not yet stable as a const fn
+
+const fn transmute_fn() -> u32 { unsafe { mem::transmute(Foo(3)) } }
```
Comment (RalfJung) [inline src/test/ui/feature-gates/feature-gate-const_transmute.rs:8]: Maybe also add a test without the `unsafe` block, and one that directly calls `core::intrinsics::transmute` (which is "accidentally" accessible on stable)? And... hm, what other combinations could we try?

**Comment by RalfJung** [comment]
> Idk who needs to sign off on this stabilization. Technically T-libs? Hypothetically just wg-const-eval? I mean this is just removing boilerplate for users for a feature already existing via unions.

Since intrinsics are involved, I think T-lang needs to sign off as well.

File: src/librustc_mir/transform/qualify_min_const_fn.rs [hunk 6]
```diff
@@ -383,9 +383,14 @@ fn check_terminator(
                         .into(),
                     ));
                 }
+
+                // HACK: This is to "unstabilize" the stabilization of the `transmute` intrinsic
```
Comment (RalfJung) [inline src/librustc_mir/transform/qualify_min_const_fn.rs:387]:
```suggestion
                // HACK: This is to "unstabilize" the `transmute` intrinsic
```
I don't know what it means to unstabilize a stabilization. ;)

File: src/test/ui/feature-gates/feature-gate-const_transmute.rs [hunk 7]
```diff
@@ -8,4 +8,10 @@ const TRANSMUTED_U32: u32 = unsafe { mem::transmute(Foo(3)) };
 const fn transmute_fn() -> u32 { unsafe { mem::transmute(Foo(3)) } }
 //~^ ERROR can only call `transmute` from const items, not `const fn`
 
+const fn transmute_fn_intrinsic() -> u32 { unsafe { std::intrinsics::transmute(Foo(3)) } }
+//~^ ERROR can only call `transmute` from const items, not `const fn`
```
Comment (RalfJung) [inline src/test/ui/feature-gates/feature-gate-const_transmute.rs]: What about outside `unsafe` blocks in a `unsafe const fn` (or maybe entirely outside `unsafe`)?

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs:387]: that's why it's in quotes. I was hoping the rest of the text would make it clear.

File: src/test/ui/feature-gates/feature-gate-const_transmute.rs (hunk 7 above)
Comment (oli-obk) [reply src/test/ui/feature-gates/feature-gate-const_transmute.rs]: that's unaffected by const fn. It's a completely different source path that checks for that, and it completely ignores the `const` modifier on functions.

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 6 above)
Comment (RalfJung) [reply src/librustc_mir/transform/qualify_min_const_fn.rs:387]: I mean it's clear what you mean, just like it's mostly still clear what you mean when there's a typo. I think with my edit/"typo" fix it's just as clear but sounds more like English. ;)

File: src/test/ui/feature-gates/feature-gate-const_transmute.rs (hunk 7 above)
Comment (RalfJung) [reply src/test/ui/feature-gates/feature-gate-const_transmute.rs]: Sure. But it's cheat to add tests for that here, isn't it?

Comment (oli-obk) [reply src/test/ui/feature-gates/feature-gate-const_transmute.rs]: :shrug: I added some

**Comment by oli-obk** [comment]
@rust-lang/lang This PR stabilizes `transmute` in const eval, but still keeps it forbidden inside `const fn`. We already do the same thing for `union` field accesses. Considering that users can already do everything `transmute` can do, but with more boilerplate and in a more error-prone way, it seems to wg-const-eval that we should just allow `transmute` wherever `union` field accesses are allowed.

I do not know of any concerns that would speak against doing this, except for impl complexity. This PR shows all the complexity needed and is fairly minimal imo. The additional code will go away once we decide to allow `transmute` and `union` field accesses in `const fn`.

**Comment by nikomatsakis** [comment]
The idea of "transmute is permitted wherever union fields are" seems reasonable to me. I'm going to go ahead and kick off a fcp merge here -- ordinarily I'd prefer a more "formalized" stabilization report, but I think that [@oli-obk's comment, plus the tests one can see in the diff here, probably suffices](https://github.com/rust-lang/rust/pull/72920#issuecomment-638306397). 

@rfcbot fcp merge

**Comment by nikomatsakis** [comment]
Discussed in triage meeting today: felt ok about this. =)

**Comment by nikomatsakis** [comment]
Briefly, in the @rust-lang/lang meeting we were wondering why unions are disallowed in const fns, though we don't think that should block progress on this issue.

**Comment by oli-obk** [comment]
There lots (and even recent) discussion on unions in const fn in https://github.com/rust-lang/rust/issues/51909

TLDR is that as of right now you can't do unsafe things in const fn, and we first need to communicate the difference in safety rules in const vs non-const contexts.

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/transform/qualify_min_const_fn.rs:399]: > In the meantime, we should check this in check_consts/validation.rs as well. Adding a structured error type for transmute and checking for it where we handle TerminatorKind::Call should be enough.

@ecstatic-morse so are you saying that should happen in this PR, or should be done some time after landing this PR but before removing `min_const_fn`?

**Comment by oli-obk** [comment]
> Discussed in triage meeting today: felt ok about this. =)

I'm guessing that we'll still need to follow the FCP procedure.

ping @pnkfelix, @cramertj, @scottmcm and @withoutboats there are checkboxes in https://github.com/rust-lang/rust/pull/72920#issuecomment-640694243 for your to click or raise concerns about.

**Comment by RalfJung** [comment]
What is the usual procedure for FCPs that went a few weeks without T-lang activity?

**Comment by nikomatsakis** [comment]
@RalfJung the usual procedure sadly is to do nothing, but I'm trying to change that. I'll try to direct some attention here,. To start:

cc @pnkfelix @cramertj @scottmcm @withoutboats

File: src/libcore/intrinsics.rs [hunk 8]
```diff
@@ -1284,7 +1284,9 @@ extern "rust-intrinsic" {
     /// }
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
-    #[rustc_const_unstable(feature = "const_transmute", issue = "53605")]
+    // NOTE: While this makes the intrinsic const stable, we have some custom code in const fn
+    // checks that prevent its use within `const fn`.
+    #[rustc_const_stable(feature = "const_transmute", since = "1.45.0")]
```
Comment (oli-obk) [inline src/libcore/intrinsics.rs:1289]: note to self, move goal post to 1.46

File: src/librustc_feature/active.rs [hunk 9]
```diff
@@ -574,6 +574,9 @@ declare_features! (
     /// No longer treat an unsafe function as an unsafe block.
     (active, unsafe_block_in_unsafe_fn, "1.45.0", Some(71668), None),
 
+    /// Alloc calling `transmute` in const fn
+    (active, const_fn_transmute, "1.45.0", Some(53605), None),
```
Comment (oli-obk) [inline src/librustc_feature/active.rs:578]:
```suggestion
    (active, const_fn_transmute, "1.46.0", Some(53605), None),
```

**Comment by oli-obk** [comment]
rebased

**Comment by RalfJung** [comment]
Awesome, let's ship this. :)
@bors r+

**Comment by Manishearth** [comment]
@bors r-

Fails https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/missing_const_for_fn/could_be_const.rs

**Comment by RalfJung** [comment]
So what changed to fix the issue? When you amend and rebase at the same time, that is impossible to tell (so it would be better if you didn't do that IMO).

**Comment by Manishearth** [comment]
He just tweaked the clippy test output, no big deal

**Comment by RalfJung** [comment]
The diff contains no file with "clippy" in its name though.

**Comment by Manishearth** [comment]
It does, last file

**Comment by RalfJung** [comment]
Oh I see. But when I use the GitHub "jump to" menu and enter "clippy", no results are shown? That's how I concluded that no such file is touched. What is GH even doing...

**Comment by Manishearth** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/74242#issuecomment-657076737

**Comment by Manishearth** [comment]
@bors r=RalfJung

actually this is okay, but https://github.com/rust-lang/rust/pull/74203 breaks it

**Comment by Manishearth** [comment]
This PR is currently in a rollup and https://github.com/rust-lang/rust/pull/72920 is not yet fixed, but if https://github.com/rust-lang/rust/pull/72920 lands first this PR will need to be updated.

**Comment by Manishearth** [comment]
Feel free to update the PR here with a new commit and let me know so I can include the update in the next rollup. Or make a separate PR.

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rust/pull/74245/commits/c8c4fd7cb17216c4771c03e521dfea073c16df8a

**Comment by RalfJung** [comment]
Somehow, this did not make the Rust 1.46 release notes, even though the relevant tag is set? Cc @rust-lang/release

**Comment by jonas-schievink** [comment]
Yes, the blog post just highlights a select few notable changes.
